2024/python/day01/part2.py

```diff
@@ -0,0 +1,32 @@
+# part2.py
+import sys
+
+def main():
+    """
+    Main function to read the input file, process the entries, and calculate the similarity score, with a metric based
+    on how often numbers in the left list show up in the right list.
+    """
+    filename = sys.argv[1]
+    try:
+        with open(filename, 'r') as file:
+            similarity = 0
+            left_list = []
+            right_list = []
+
+            for line in file:
+                entries = line.strip().split() # strips newlines and spilts by spaces
+                if entries != []:
+                    left_list.append(entries[0])
+                    right_list.append(entries[1])
+                
+            file.close()
+
+        for id in left_list:
+            similarity += int(id) * right_list.count(id)
```

Comment:
what's the time complexity of this loop? and is there a way to make it better?

Reply:
Original time complexity was `O(nxm)` time complexity because we are scanning the entire `right_list` for each element in `left_list`. 

Counter is implemented in C and does the same work. 

Time Complexity: `right_list` - `O(m)`
```suggestion
from collections import Counter

right_counter = Counter(right_list)
similarity = sum(int(id) * right_counter[id] for id in left_list)
```

Time Complexity: Similarity calculation - `O(n)`
```suggestion
right_counts = {}
for item in right_list:
    right_counts[item] = right_counts.get(item, 0) + 1

similarity = sum(int(id) * right_counts.get(id, 0) for id in left_list)
```

Time Complexity:` O(n)`
```suggestion
right_counts = {}
for item in right_list:
    right_counts[item] = right_counts.get(item, 0) + 1

for id in left_list:
    similarity += int(id) * right_counts.get(id, 0)
```